### ISSUE TITLE
Rework connection initiation sequence.

### ISSUE THREAD

**Issue by mondalaci** [UltimateHackingKeyboard/firmware#983]
Both halves are powered only via SWD, which is not a legitimate use case, but I'd still expect both batteries to show.

![IMG20240610185132](https://github.com/UltimateHackingKeyboard/firmware/assets/300402/08ab68cc-ce8d-4aad-8a56-cbd3584d533f)

When connecting UART, both values show up.

**Comment by kareltucek** [comment]
> Both halves are powered only via SWD,

But the batteries are connected, right?

**Comment by mondalaci** [comment]
Yes, although I get "0- 0-" when connecting the left half via UART, even if the left battery is not present.

**Comment by kareltucek** [comment]
Can you reproduce it consistently or does it happen only time to time?

Allright, please run `uhk charger` and see whether it produces correct results.

**Comment by mondalaci** [comment]
Yes, I can consistently reproduce it.

* left: `CHARGER_EN: 1 | STAT: 1 Battery is present: 1, charging: 0, at 2950mV (0%)`
* right: `CHARGER_EN: 1 | STAT: 1 Battery is present: 1, charging: 0, at 2884mV (0%)`

**Comment by mondalaci** [comment]
I can reproduce this issue even when the halves are connected with the bridge cable after rebooting the right half. Afterward, if I reconnect the bridge cable, the left battery appears.

I'm suspending the testing of UltimateHackingKeyboard/firmware#875 as this issue makes testing hard.

**Comment by kareltucek** [comment]
🤦

I see. I guess the right half reboots and starts pinging uart under 2 seconds, which means that left half never notices that the connection was broken. (Left half resets its dirty flags whenever connection parameters to the right half change.)

**Comment by kareltucek** [comment]
Well, I am offering this branch: https://github.com/UltimateHackingKeyboard/firmware-uhk80/pull/139

I originally wrote it as a hackfix for the bluetooth troubles and didn't bother to post it as a PR as long as there is a chance that we manage to resolve the NUS initializations problems cleanly, but in the light of the bridge cable problem, I am wondering if we should actually employ it.

**Comment by kareltucek** [comment]
As to a clean solution to the UART problem, The only thing that comes to mind is to make UART not start pinging until at least 2 seconds (or more generally the UART_TIMEOUT) after boot. Should I employ it?

**Comment by mondalaci** [comment]
Your UART PR doesn't seem very robust. I'd prefer peers (such as the left half, right half, and dongle) to explicitly notify each other after reboot to resync properties.

**Comment by kareltucek** [comment]
Well, that's what https://github.com/UltimateHackingKeyboard/firmware-uhk80/pull/139 is basically doing. I can extend it to work in all directions.

Should I?

**Comment by kareltucek** [comment]
Or you mean doing this only on UART level?

**Comment by mondalaci** [comment]
I think UltimateHackingKeyboard/firmware-uhk80#139 is a robust, general solution, so using it for both NUS and UART is fine.

Should I close this issue? It seems to be fully resolved, but you may want to take further steps depending on Nordic's suggestion regarding the bidirectional establishment callback of the NUS connection.

**Comment by kareltucek** [comment]
Lets keep it open. UltimateHackingKeyboard/firmware-uhk80#139 while working, I think we should probably rework it for it to be fully general and more robust.

**Comment by mondalaci** [comment]
Since UltimateHackingKeyboard/firmware-uhk80#159, the LEDs have sometimes been all over the place. I'm afraid this won't fly well during the FCC certification, so please make fixing this a priority.

**Comment by kareltucek** [comment]
@mondalaci I'm afraid I still have no solution for NUS initialization.

**Comment by kareltucek** [comment]
Also, can you provide more details about "LEDs have sometimes been all over the place" (maybe in a separate issue)?

**Comment by mondalaci** [comment]
Regarding NUS initialization, are we waiting for Nordic? If not, what's the sticking point?

Regarding the LED issue, if I restart the right half, I get this on the left half:

![IMG20240618000819](https://github.com/UltimateHackingKeyboard/firmware-uhk80/assets/300402/54cb2440-1ce7-417a-bf83-e248e2d1965d)

**Comment by kareltucek** [comment]
Yes, for nordic. I have got no answer in the last 6 days. I have bumped the thread today... https://devzone.nordicsemi.com/f/nordic-q-a/111767/when-is-nus-ready

As for leds, I can't reproduce that. Are you sure your left half is running the same firmware version? (I have done some changes in the protocol in 159, so it may be receiving garbled data. Although that should definitely not exhibit after restart only.)

**Comment by kareltucek** [comment]
Nevermind, I think I have found an error in my code. 

I don't understand why it has not exhibited here though.

**Comment by mondalaci** [comment]
Sorry, my bad! I was on the wrong branch, I've reflashed the halves, and I can't reproduce the issue since then.

**Comment by kareltucek** [comment]
@mondalaci are the halves "paired and bonded"? https://devzone.nordicsemi.com/f/nordic-q-a/111767/when-is-nus-ready/489167

**Comment by mondalaci** [comment]
Yes.

If you run `settings list` in the shell, you can see that the halves have each others' security keys, such as `bt/keys/e7f55d7c82351`.

This is akin to pairing the phone and not making the phone forget the device. It's worth noting that even if a phone forgets the UHK 80, the UHK 80 will still have the phone's security keys, which we'll have to delete from the settings partition explicitly. Otherwise, it can eventually overflow.

**Comment by kareltucek** [comment]
This seems no longer relevant.
